app/src/main/java/io/gnosis/safe/ui/transaction/TransactionsViewModel.kt

```diff
@@ -101,14 +454,119 @@ class TransactionsViewModel
+        override val status: TransactionStatus,
+        @StringRes val statusText: Int,
+        @ColorRes val statusColorRes: Int,
+        val dateTimeText: String,
+        val settingNameText: String,
+        val alpha: Float
+    ) : TransactionView(status)
+
+    data class SettingsChangeQueued(
+        override val status: TransactionStatus,
+        @StringRes val statusText: Int,
+        @ColorRes val statusColorRes: Int,
+        val dateTimeText: String,
+        val settingNameText: String,
+        val confirmations: Int,
+        val threshold: Int,
+        @ColorRes val confirmationsTextColor: Int,
+        @DrawableRes val confirmationsIcon: Int,
+        val nonce: String
+    ) : TransactionView(status)
+
+    data class ChangeMastercopy(
+        override val status: TransactionStatus,
+        @StringRes val statusText: Int,
+        @ColorRes val statusColorRes: Int,
+        val dateTimeText: String,
+        val address: Solidity.Address?,
+        val version: String,
+        val alpha: Float,
+        @StringRes val label: Int,
+        val visibilityVersion: Int = View.VISIBLE,
+        val visibilityEllipsizedAddress: Int = View.VISIBLE,
+        val visibilityModuleAddress: Int = View.GONE
+
+    ) : TransactionView(status)
+
+    data class ChangeMastercopyQueued(
+        override val status: TransactionStatus,
+        @StringRes val statusText: Int,
+        @ColorRes val statusColorRes: Int,
+        val dateTimeText: String,
+        val confirmations: Int,
+        val threshold: Int,
+        @ColorRes val confirmationsTextColor: Int,
+        @DrawableRes val confirmationsIcon: Int,
+        val nonce: String,
+        val address: Solidity.Address?,
+        val version: String,
+        @StringRes val label: Int,
+        val visibilityVersion: Int = View.VISIBLE,
+        val visibilityEllipsizedAddress: Int = View.VISIBLE,
+        val visibilityModuleAddress: Int = View.GONE
+    ) : TransactionView(status)
+
+    data class CustomTransaction(
+        override val status: TransactionStatus,
+        @StringRes val statusText: Int,
+        @ColorRes val statusColorRes: Int,
+        val dateTimeText: String,
+        val address: Solidity.Address,
+        val dataSizeText: String,
+        val amountText: String,
+        @ColorRes val amountColor: Int,
+        val alpha: Float
+    ) : TransactionView(status)
+
+    data class CustomTransactionQueued(
+        override val status: TransactionStatus,
+        @StringRes val statusText: Int,
+        @ColorRes val statusColorRes: Int,
+        val dateTimeText: String,
+        val address: Solidity.Address,
+        val confirmations: Int,
+        val threshold: Int,
+        @ColorRes val confirmationsTextColor: Int,
+        @DrawableRes val confirmationsIcon: Int,
+        val dataSizeText: String,
+        val amountText: String,
+        @ColorRes val amountColor: Int,
+        val nonce: String
+    ) : TransactionView(status)
```

Comment:
Oh man, this really blew up in terms of complexity. Why did you opt out from simply passing the `Transaction.SettingsChange` or `Transaction.Transfer` or `Transaction.Custom` upwards? These UI assets would make total sense to resolve in the `viewHolder`, actually you have them requested and used in the same place. Here in the `TransactionsViewModel` feels to me that it loses a bit of visibility to where all of this is used.

Reply:
There is a lot of UI specific stuff that doesn't need to be defined by the viewModel. I think a lot of the logic that you defined here would be easier to digest if it were placed in the respective viewHolder, as a lot of it, is only applicable to the specific layout the viewHolder binds to.

Reply:
I think it is ok to have a viewModel/viewHolder pair defined for each cell, but as it is it feels a bit like a monolithical solution provided uniquely by the `TransactionsViewModel` which is a bit detrimental to readability. I think maybe it would be a good idea to address this in the future. As this, I suppose, works already. Then we could plan for a bit of a refactor/cleanup, if you would agree?

Reply:
The viewHolder classes should not have any logic. The logic should be in a class that can be unit tested. That is why I decided to do the conversion in the viewModel. We can certainly moved it somewhere else. But the TransactionView classes should be dumb data containers.

Reply:
ViewHolders don't need to be data classes. I guess thats what you're referring to? Let's do a refactoring! As a separate PR. We could probably get rid of  a lot of duplication. Maybe we don't need independent classes for queued and history txs, for example.

Reply:
Absolutely. I totally agree with you on what the viewHolders should do. My point was more towards the fact that the `TransactionView` classes became kind of a full view descriptor; holding info about variables that are too close to the view layer, like alpha, colours and string asset ids. That's a little too fine grained in my opinion.